RecoMET/METAlgorithms/src/CSCHaloAlgo.cc

```diff
@@ -757,44 +639,695 @@ math::XYZPoint CSCHaloAlgo::getPosition(const DetId &id, reco::Vertex::Point vtx
+  //Separate conditions targeting IT and OT beam halos
+  bool halofound = false;
+
+  //For OT beam halos, just require enough crystals with large T
+    for(size_t i = 0; i <haloclustercands.size(); i++){
+      if(haloclustercands[i].GetSeedEt()<20)continue;
+      if(haloclustercands[i].GetSeedTime()<0.5)continue;
+      if(haloclustercands[i].GetNbLateCrystals()-haloclustercands[i].GetNbEarlyCrystals() <2)continue;
+      
+      //The use of time information does not allow this method to work at HLT
+      if(ishlt)continue;
+      halofound =true;
+      edm::RefVector<EcalRecHitCollection> bhrhcandidates = haloclustercands[i].GetBeamHaloRecHitsCandidates();
+      if(halofound)AddtoBeamHaloEBEERechits(bhrhcandidates, thehalodata,false);
+      
+    }
+
+    //For IT beam halos, fakes from collisions are higher => require the cluster size to be small. 
+    //Only halos with R>100 cm are considered here.
+    //For lower values, the time difference with particles from collisions is too small
+    //IT outgoing beam halos that interact in EE at low R is probably the most difficult category to deal with: 
+    //Their signature is very close to the one of photon from collisions (similar cluster shape and timing)
+    for(size_t i = 0; i <haloclustercands.size(); i++){
+      if(haloclustercands[i].GetSeedEt()<20)continue;
+      if(haloclustercands[i].GetSeedR()<100)continue;
+      if(haloclustercands[i].GetTimeDiscriminator()<1) continue; 
+      if(haloclustercands[i].GetClusterSize()<2) continue;
+      if(haloclustercands[i].GetClusterSize()>4) continue;
+      
+      //The use of time information does not allow this method to work at HLT
+      if(ishlt)continue;
+      halofound =true;
+      edm::RefVector<EcalRecHitCollection> bhrhcandidates = haloclustercands[i].GetBeamHaloRecHitsCandidates();
+      if(halofound) AddtoBeamHaloEBEERechits(bhrhcandidates, thehalodata,false);
+
+
+    }
+    
+    return halofound;
+
+}
+
+
+
+bool CSCHaloAlgo::HBClusterShapeandTimeStudy(reco::CSCHaloData & thehalodata,std::vector<HaloClusterCandidateHB>haloclustercands, bool ishlt){
+  //Conditions on the central strip size in eta.
+  //For low size, extra conditions on seed et, isolation and cluster timing 
+  //Here we target both IT and OT beam halo. Two separate discriminators were built for the two cases.
+  bool halofound = false;
+  for(size_t i = 0; i <haloclustercands.size(); i++){
+    if(haloclustercands[i].GetSeedEt()<10)continue;
+    if(haloclustercands[i].GetNbTowersInEta()<3) continue;
+    //Isolation criteria for very short eta strips
+    if(haloclustercands[i].GetNbTowersInEta()==3 && (haloclustercands[i].GetEtStripPhiSeedPlus1()>0.1 || haloclustercands[i].GetEtStripPhiSeedMinus1()>0.1) ) continue;
+    if(haloclustercands[i].GetNbTowersInEta()<=5 && (haloclustercands[i].GetEtStripPhiSeedPlus1()>0.1 && haloclustercands[i].GetEtStripPhiSeedMinus1()>0.1) ) continue;
```

Comment:
the c-style for loop  can be changed to range loop ``` for (hcand : haloclustercands ) { ``` 
This will reduce amount of text and potential multiple calls to  ``` haloclustercands[i] ```